[Salesforce] Bulk Upsert

nick-Ag opened the pull request:
<!-- Hello and thank you for contributing to Segment action-destinations! -->

<!-- Before opening your pull request, make sure you have added and ran unit
     tests and tested your change locally. Refer to our testing
     documentation for more information: https://github.com/segmentio/action-destinations/blob/main/docs/testing.md -->

<!-- If you have questions or issues please open a new issue or create a new discussion
     post in Github. -->

This PR adds support for batching to Salesforce. It includes a new "Bulk Upsert" operation which is included in every action. This is an initial build that will only be usable by internal Segment users. More details on that can be found in the [(Segment Internal) Bulk API Mini-SDD](https://paper.dropbox.com/doc/Mini-SDD-Salesforce-Bulk-API--BhoQssc8VEjaaGVvnkxwnG0EAg-OW8Dn2fzQRDPlCa8trLLG).

Corresponding App PR which locks this to Segment Internal: https://github.com/segmentio/app/pull/13309


## Testing
Deployed successfully to stage. Batches of events arrive in the `performBatch` method, and are correctly delivered downstream to Salesforce using the bulk job flow.

Some pictures:
<img width="1511" alt="Screen Shot 2022-05-18 at 4 28 06 PM" src="https://user-images.githubusercontent.com/27820201/169172314-d52eb189-5ac7-4b21-8231-2f34181163dc.png">

<img width="1305" alt="Screen Shot 2022-05-18 at 4 27 54 PM" src="https://user-images.githubusercontent.com/27820201/169172304-5a38a5cc-297a-4f40-842b-e3f6e025a81f.png">

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-object-to-shape.ts:
```diff
@@ -95,3 +95,13 @@ export const mapObjectToShape = (payload: GenericPayload, sobject: string): Gene
+
+// Our key names are in snake case, but Salesforce's field names are in camel case.
+// I.E. 'last_name' is our key, but 'LastName' is the Salesforce field name.
+export const snakeCaseToCamelCase = (key: string): string => {
```
I believe we have lodash in this repo so can potentially use `_.camelCase` https://lodash.com/docs/#camelCase

nick-Ag commented on packages/destination-actions/src/destinations/salesforce/opportunity/index.ts:
```diff
@@ -60,6 +60,17 @@ const action: ActionDefinition<Settings, Payload> = {
+  performBatch: async (request, { settings, payload }) => {
+    const sf: Salesforce = new Salesforce(settings.instanceUrl, request)
+
+    if (payload[0].operation === 'bulkUpsert') {
```
An open question I have: Right now I'm settling for checking the first item in Payload[]. I would like to loop thru the array and return errors only for those missing required fields, and proceed with the other events. Not sure if that's currently possible

dlasky replied:
If they are required and missing I believe the validation step will have already discarded them--see the batching unit tests for an example there.

mericsson commented on packages/destination-actions/src/destinations/salesforce/__tests__/sf-operations.test.ts:
```diff
@@ -254,4 +255,66 @@ describe('Salesforce', () => {
     })
   })
+
+  describe.only('Bulk Operations', () => {
```
```suggestion
  describe('Bulk Operations', () => {
```

mericsson commented on packages/destination-actions/src/destinations/salesforce/account/index.ts:
```diff
@@ -185,6 +185,17 @@ const action: ActionDefinition<Settings, Payload> = {
       return await sf.upsertRecord(payload, 'Account')
     }
+  },
+  performBatch: async (request, { settings, payload }) => {
```
I wonder if adding this will cause existing traffic to start hitting `performBatch`? Should we add fallback here to single `perform` if `bulkUpsert` is not enabled?

mericsson replied:
(And same for all other added `performBatch` calls)

nick-Ag replied:
I think that would be possible if we somehow get an array to arrive in [this code block](https://github.com/segmentio/integrations/blob/master/createActionDestination/index.js#L210) for existing traffic. But if I understand correct that's only possible if users enable `enable_batching` (will be locked to internal for now). But I will test this in stage Monday (pushed that code to the Opprotunity action for testing)

mericsson replied:
I think as soon as `performBatch` is included as a function and pushed with the register update it will start trying to send batch events.

https://github.com/segmentio/action-destinations/blob/160cc4f31908ee08820cd8ea5d4e429e2dee7465/packages/cli-internal/src/commands/push.ts#L171-L184

May need to patch the CLI to consider an additional field name like `batch_enabled` on the action definition. What do you think @dlasky ?

dlasky replied:
thats what the code linked above does, that adds a user controllable MAPPING for enable_batching which shows up as a boolean in the mapping config.

mericsson commented on packages/destination-actions/src/destinations/salesforce/cases/index.ts:
```diff
@@ -33,6 +33,13 @@ const action: ActionDefinition<Settings, Payload> = {
+    const sf: Salesforce = new Salesforce(settings.instanceUrl, request)
+
+    if (payload[0].operation === 'bulkUpsert') {
+      return await sf.bulkUpsert(payload, 'Case')
```
I think cleaner / safer to define a static `const` in this file and re-use.

```js
const objectName = 'Case'
```

and 

```js
return await sf.bulkUpsert(payload, objectName)
```

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-operations.ts:
```diff
@@ -1,8 +1,13 @@
+import { mapObjectToShape, snakeCaseToCamelCase } from './sf-object-to-shape'
 
 export const API_VERSION = 'v53.0'
+const isSettingsKey = new Map<string, boolean>([
+  ['operation', true],
+  ['traits', true],
+  ['externalIdFieldName', true]
+])
```
Nitpick, can use a `Set` instead.

```suggestion
const isSettingsKey = new Set<string>(['operation', 'traits', 'externalIdFieldName'])
```

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-operations.ts:
```diff
@@ -68,6 +77,110 @@ export default class Salesforce {
+    })
+  }
+
+  private buildCSVData = (payloads: GenericPayload[], externalIdFieldName: string): string => {
```
I thin we should add some unit tests for these `build*` functions.

mericsson replied:
Probably good idea to pull it out into its own file too for clarity e.g. csv-util.ts

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-operations.ts:
```diff
@@ -68,6 +77,110 @@ export default class Salesforce {
+  private buildHeader = (payloads: GenericPayload[], externalIdFieldName: string): string => {
+    let header = ''
+    Object.keys(payloads[0]).forEach((key) => {
+      console.log('CSV: ', key)
```
```suggestion
```

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-operations.ts:
```diff
@@ -68,6 +77,110 @@ export default class Salesforce {
+
+    Object.entries(payload).forEach(([key, value]) => {
+      if (!isSettingsKey.get(key)) {
+        row += `"${value}",`
```
May need to additionally escape `,` values that may reside inside of `value`.

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-operations.ts:
```diff
@@ -68,6 +77,110 @@ export default class Salesforce {
+
+    // Iterate over the keys in payload that map to a SF object field
+    // Add each value to the row, separated by a comma
+    // The last value in the row should be the externalIDField value
```
Could you add more info here about why the last value should be `externalIDField`? Not clear to me.

mericsson replied:
I suppose that's because it's last in the header? Curious if there is a reason it is last or first?

nick-Ag replied:
I could put the `externalIDField` anywhere in the header, I put it last since it makes handling the last comma super simple. If I know the last item is always externalIDField, I just write that at the end without a comma

nick-Ag replied:
(added to in code comment)

mericsson replied:
👍

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-operations.ts:
```diff
@@ -68,6 +77,110 @@ export default class Salesforce {
+    Object.keys(payloads[0]).forEach((key) => {
+      console.log('CSV: ', key)
+      if (!isSettingsKey.get(key)) {
+        header += `"${snakeCaseToCamelCase(key)}",`
```
May need to escape or error if `key` contains `,` already.

nick-Ag replied:
I've added this validation to `customFields` only. Since we control the `key` value for the other keys and don't include any invalid names

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-properties.ts:
```diff
@@ -10,7 +10,8 @@ export const operation: InputField = {
     { label: 'Update', value: 'update' },
-    { label: 'Upsert', value: 'upsert' }
+    { label: 'Upsert', value: 'upsert' },
+    { label: 'Bulk Upsert', value: 'bulkUpsert' }
```
This will be hidden in the UI for now? Is that correct?

nick-Ag commented:
Addressed comments and refactored to use a new `sf-utils.ts` file for CSV generation methods. Also supports `customFields` in the header now.
A couple TODOs:
* Support incomplete data for CSV generation. Right now if payload[0] has a name value and payload[1] doesn't have a name value the CSV will be incorrectly formatted.
* Test `performBatch` hijacking existing data meant to arrive in the `perform` method in stage.
* Escape double quotes inside CSV values

mericsson commented on packages/destination-actions/src/destinations/salesforce/sf-utils.ts:
```diff
@@ -0,0 +1,86 @@
+  return header
+}
+
+// Our key names are in snake case, but Salesforce's field names are in Pascal case.
```
Sounds good. Just an fyi -- Salesforce API for field names is case insensitive. So you don't strictly need PascalCase. (But doesn't hurt -- it is just a convention).

mericsson replied:
(e.g. You could just get away with `camelCase` only)

nick-Ag commented on packages/destination-actions/src/destinations/salesforce/sf-utils.ts:
```diff
@@ -0,0 +1,143 @@
+ * @param externalIdFieldName The name of the field that contains the external ID.
+ * @returns The complete CSV to send to Salesforce.
+ */
+export const buildCSVData = (payloads: GenericPayload[], externalIdFieldName: string): string => {
```
The entry point for CSV generation. Supporting incomplete payload data ended up being complex, so please leave plenty of feedback on this

mericsson replied:
I suppose the problem is that inside of `performBatch` we just get a list of payloads and not any metadata about the mappings or what the full event could be? Ideally we would just get the mappings that the customer configured in the UI and use that as the header, right? If so, could be a follow-up to provide that in some way.

nick-Ag replied:
The header ordering doesn't matter no. I'll go ahead and address this in a follow-up :+1:

nick-Ag commented:
Hey team, this new build includes supporting incomplete Payload data. I've tested in stage and locally, plus added new unit tests and the functionality looks good. That being said this adds complexity to the CSV generation, so any feedback is greatly appreciated!

rhall-twilio commented on packages/destination-actions/src/destinations/salesforce/__tests__/sf-operations.test.ts:
```diff
@@ -254,4 +255,130 @@ describe('Salesforce', () => {
+      }
+    ]
+
+    it('should correctly upsert a batch of records', async () => {
```
I would expect testcases to have at least one expect() statement to provide a passing validation criteria. 

I think right now we are assuming that sf.bulkUpsert not throwing means that the code did what was intended, but that might not be necessarily true. e.g. is there a way to confirm that the nock endpoints got hit? or does the test automatically fail if they don't?

rhall-twilio replied:
Maybe https://michaelheap.com/nock-all-mocks-used/

nick-Ag replied:
Enabled this and the case when a nock interceptor is not triggered throws a much more readable error than before. Thanks for the suggestion!

rhall-twilio commented on packages/destination-actions/src/destinations/salesforce/account/index.ts:
```diff
@@ -170,20 +173,31 @@ const action: ActionDefinition<Settings, Payload> = {
+  performBatch: async (request, { settings, payload }) => {
+    const sf: Salesforce = new Salesforce(settings.instanceUrl, request)
+
+    if (payload[0].operation === 'bulkUpsert') {
```
What happens if the operation is not bulkUpsert and we hit performBatch?

mericsson replied:
I think may be a good idea to add an so it doesn't "silently fail" just in case someone misconfigures it.

```js
else {
//  throw IntegrationError ...
}
```

rhall-twilio commented on packages/destination-actions/src/destinations/salesforce/sf-utils.ts:
```diff
@@ -0,0 +1,143 @@
+      if (key === 'customFields') {
+        const customFields = value as object
+        Object.entries(customFields).forEach(([customFieldName, customFieldValue]) => {
+          const fieldErr = validateHeaderField(customFieldName)
```
Is there a reason the validation method doesn't throw the exception? It's an interesting pattern to rely on the client to throw the error, particularly if we are not aggregating them.

mericsson replied:
+1 I think should just `throw` from `validateHeaderField`. Would simplify this a little.

mericsson commented on packages/destination-actions/src/destinations/salesforce/opportunity/index.ts:
```diff
@@ -45,20 +48,34 @@ const action: ActionDefinition<Settings, Payload> = {
+
+      return await sf.bulkUpsert(payload, OBJECT_NAME)
+    } else {
+      const errorMsg = 'Bulk Upsert action must be used with batching'
+      throw new IntegrationError(errorMsg, errorMsg, 400)
```
Not blocking but I guess the second argument should is an optional code.

```suggestion
      const errorMsg = 'Bulk Upsert action must be used with batching'
      throw new IntegrationError(errorMsg, 'Bad Request', 400)
```

https://github.com/segmentio/action-destinations/blob/4a53c8bbb3a07d4b4aa0b391d377ef1ccd58ca3f/packages/destination-actions/src/destinations/slack/postToChannel/index.ts#L50

mericsson replied:
(As a larger follow-up we should make `IntegrationError` easier to get right. I see a lot of different usage of it across action-destinations)
